File: boj/1935.ml
let input () =
  let vars = read_int () in
  let exp = read_line () in
  let var_map = Hashtbl.create vars in
  for i = 1 to vars do
    let value = read_int () |> float_of_int in
    let var = i + 65 - 1 |> Char.chr in
    Hashtbl.add var_map var value
  done;
  exp, var_map
;;

let ops = [ '+', ( +. ); '-', ( -. ); '*', ( *. ); '/', ( /. ) ]

let eval exp var_map =
  let stack : float Stack.t = Stack.create () in
  let eval char =
    match char with
    | 'A' .. 'Z' ->
      let value = Hashtbl.find var_map char in
      Stack.push value stack
    | ('+' | '-' | '*' | '/') as op ->
      let v1 = Stack.pop stack in
      let v0 = Stack.pop stack in
      let operator = List.assoc op ops in
      let result = operator v0 v1 in
      Stack.push result stack
    | _ -> failwith "invalid char"
  in
  String.iter eval exp;
  Stack.pop stack
;;

let () =
  let exp, var_map = input () in
  let result = eval exp var_map in
  Printf.printf "%.2f" result
;;

File: refs/1935.py
import sys
def f():
    return sys.stdin.readline().strip()

total = int(f())
exp = f()
var_map = {}
for i in range(total):
    var_map[chr(65+i)] = int(f())

op_map = {
    '*': lambda x, y: x * y,
    '+': lambda x, y: x + y,
    '-': lambda x, y: x - y,
    '/': lambda x, y: x / y,
}

stack = []
for e in exp:
    if e.isalpha():
        stack.append(var_map[e])
    else:
        # e is an operator
        v2, v1 = stack.pop(), stack.pop()
        v = op_map[e](v1, v2)
        stack.append(v)

print(f'{stack[0]:.2f}')
